src/encoding/json/encode.go

```diff
@@ -312,6 +312,14 @@ func newEncodeState() *encodeState {
 	return &encodeState{ptrSeen: make(map[any]struct{})}
 }
 
+func cacheEncodeState(e *encodeState) {
+	// caching large objects is not memory efficient.
+	if e.Buffer.Cap() > 32768 {
```

Comment:
Some limit seems like a great idea, but how did you come up with this specific number?

Reply:
I am open to any numbers <= 1MiB, which is good enough for apiserver issue that I am trying to fix

Reply:
I recommend doing some kind of search process to recommend an optimal value or a tighter range. E.g. binary search through possible values to find the lowest and highest ones that fix your problem.

I suspect your particular test won't be able to narrow this down too much, you'll need a more realistic object size distribution. I'm not sure where to get that from.

Reply:
Actually, I just read this code more and I agree that caching large objects is going to be horribly memory inefficient because they can be reused to store small objects, which will retain all that memory for little reason. I'm not sure where a good threshold is but it may be even smaller than this.